File: web/lib/workflow-variable-validation.ts
import type { WorkflowDefinitionPreflightIssue, WorkflowDetail } from "@/lib/get-workflows";

export type WorkflowVariableValidationIssue = WorkflowDefinitionPreflightIssue & {
  category: "variables";
};

export function buildWorkflowVariableValidationIssues(
  definition: WorkflowDetail["definition"]
): WorkflowVariableValidationIssue[] {
  const variables = Array.isArray(definition.variables) ? definition.variables : [];
  const issues: WorkflowVariableValidationIssue[] = [];
  const seenNames = new Map<string, number[]>();

  variables.forEach((variable, index) => {
    const rawName = typeof variable?.name === "string" ? variable.name : "";
    const normalizedName = rawName.trim();
    if (!normalizedName) {
      issues.push({
        category: "variables",
        message: `Variable ${index + 1} 的变量名不能为空。`,
        path: `variables.${index}.name`,
        field: "name"
      });
      return;
    }

    const indexes = seenNames.get(normalizedName) ?? [];
    indexes.push(index);
    seenNames.set(normalizedName, indexes);
  });

  for (const [name, indexes] of seenNames.entries()) {
    if (indexes.length <= 1) {
      continue;
    }
    indexes.forEach((index) => {
      issues.push({
        category: "variables",
        message: `变量名 ${name} 重复，保存前需要去重。`,
        path: `variables.${index}.name`,
        field: "name"
      });
    });
  }

  return issues;
}

File: api/app/services/workflow_variable_validation.py
from __future__ import annotations

from typing import Any


def collect_invalid_workflow_variables(
    definition: dict[str, Any] | None,
) -> list[dict[str, str]]:
    if not isinstance(definition, dict):
        return []

    variables = definition.get("variables")
    if not isinstance(variables, list):
        return []

    issues: list[dict[str, str]] = []
    seen_names: dict[str, list[int]] = {}
    for index, variable in enumerate(variables):
        if not isinstance(variable, dict):
            continue
        raw_name = variable.get("name")
        normalized_name = raw_name.strip() if isinstance(raw_name, str) else ""
        if not normalized_name:
            issues.append(
                {
                    "message": f"Variable {index + 1} uses an empty name.",
                    "path": f"variables.{index}.name",
                    "field": "name",
                }
            )
            continue
        seen_names.setdefault(normalized_name, []).append(index)

    for name, indexes in seen_names.items():
        if len(indexes) <= 1:
            continue
        for index in indexes:
            issues.append(
                {
                    "message": f"Variable name '{name}' is duplicated in the workflow definition.",
                    "path": f"variables.{index}.name",
                    "field": "name",
                }
            )

    return issues
